pkg/media/oggwriter/oggwriter.go

```diff
@@ -0,0 +1,248 @@
+	binary.LittleEndian.PutUint16(oggIDHeader[10:], defaultPreSkip) // pre-skip
+	binary.LittleEndian.PutUint32(oggIDHeader[12:], i.sampleRate)   // original sample rate, any valid sample e.g 48000
+	binary.LittleEndian.PutUint16(oggIDHeader[16:], 0)              // output gain
+	oggIDHeader[18] = 0                                             // channel map 0 = one stream: mono or stereo
+
+	// Reference: https://tools.ietf.org/html/rfc7845.html#page-6
+	// RFC specifies that the ID Header page should have a granule position of 0 and a Header Type set to 2 (StartOfStream)
+	data := i.createPage(oggIDHeader, pageHeaderTypeBeginningOfStream, 0, i.pageIndex)
+	if err := i.writeToStream(data); err != nil {
+		return err
+	}
+	i.pageIndex++
+
+	// Comment Header
+	oggCommentHeader := make([]byte, 21)
+	copy(oggCommentHeader[0:], []byte("OpusTags"))          // Magic Signature 'OpusTags'
+	binary.LittleEndian.PutUint32(oggCommentHeader[8:], 5)  // Vendor Length
+	copy(oggCommentHeader[12:], []byte("pion"))             // Vendor name 'pion'
+	binary.LittleEndian.PutUint32(oggCommentHeader[17:], 0) // User Comment List Length
+
+	// RFC specifies that the page where the CommentHeader completes should have a granule position of 0
+	data = i.createPage(oggCommentHeader, pageHeaderTypeContinuationOfStream, 0, i.pageIndex)
+	if err := i.writeToStream(data); err != nil {
+		return err
+	}
+	i.pageIndex++
+
+	return nil
+}
+
+const (
+	pageHeaderSize = 27
+)
+
+func (i *OggWriter) createPage(payload []uint8, headerType uint8, granulePos uint64, pageIndex uint32) []byte {
+	i.lastPayloadSize = len(payload)
+	page := make([]byte, pageHeaderSize+1+i.lastPayloadSize)
+
+	copy(page[0:], []byte("OggS"))                      // page headers starts with 'OggS'
+	page[4] = 0                                         // Version
+	page[5] = headerType                                // 1 = continuation, 2 = beginning of stream, 4 = end of stream
+	binary.LittleEndian.PutUint64(page[6:], granulePos) // granule position
+	binary.LittleEndian.PutUint32(page[14:], i.serial)  // Bitstream serial number
+	binary.LittleEndian.PutUint32(page[18:], pageIndex) // Page sequence number
+	page[26] = 1                                        // Number of segments in page, giving always 1 segment
+	page[27] = uint8(i.lastPayloadSize)                 // Segment Table inserting at 27th position since page header length is 27
+	copy(page[28:], payload)                            // inserting at 28th since Segment Table(1) + header length(27)
+
+	var checksum uint32
+	for index := range page {
+		checksum = (checksum << 8) ^ i.checksumTable[byte(checksum>>24)^page[index]]
+	}
+	binary.LittleEndian.PutUint32(page[22:], checksum) // Checksum - generating for page data and inserting at 22th position into 32 bits
+
+	return page
+}
+
+// WriteRTP adds a new packet and writes the appropriate headers for it
+func (i *OggWriter) WriteRTP(packet *rtp.Packet) error {
```

Comment:
Something I think it may be worth looking at (but I know nothing about this): it seems to me (by just reading the write-to-disk test), this WriteRTP method will be called in a loop and this function calls createPage which will alocate the page. This could be easily avoided by defining a buffer before the loop and using it in every iteration. Maybe is not worthy though, I'm talking by just reading at the code, don't really know much about the use case.

Reply:
Yea I don't know either. IMO the optimization isn't worth it (yet) just because we haven't measured.

I don't think the `oggwriter` is really going to see much usage besides examples either. More just a fun way for people to see it work